main/wallet.js

```diff
@@ -0,0 +1,435 @@
+    query StateReplay($cid: String!) {
+      stateReplay(cid: $cid) {
+        receipt {
+          return
+          exitCode
+          gasUsed
+        }
+        executionTrace {
+          executionTrace
+        }
+      }
+    }
+  `
+  const variables = { cid }
+  /** @type {{stateReplay: GQLStateReplay}} */
+  const { stateReplay } = await request(url, query, variables)
+  return stateReplay
+}
+
+/**
+ * @param {number} height
+ * @returns Promise<GQLTipset>
+ */
+async function getTipset (height) {
+  const query = gql`
+    query Tipset($height: Uint64!) {
+      tipset(height: $height) {
+        minTimestamp
+      }
+    }
+  `
+  const variables = { height }
+  const { tipset } = await request(url, query, variables)
+  return tipset
+}
+
+/**
+ * @param {string} address
+ * @returns Promise<GQLMessage[]>
+ */
+async function getMessages (address) {
+  const query = gql`
+    query Messages($address: String!, $limit: Int!, $offset: Int!) {
+      messages(address: $address, limit: $limit, offset: $offset) {
+        cid
+        to {
+          robust
+        }
+        from {
+          robust
+        }
+        nonce
+        height
+        method
+        params
+        value
+      }
+    }
+  `
+  const variables = {
+    address,
+    limit: 100,
+    offset: 0
+  }
+  /** @type {{messages: GQLMessage[]}} */
+  const { messages = [] } = await request(url, query, variables)
+  return messages
+}
+
+/**
+ * @returns {Promise<void>}
+ */
+async function updateTransactions () {
+  console.log(new Date(), 'updateTransactions')
+
+  // Load messages
+  const messages = await getMessages(address)
+
+  // Convert messages to transactions (loading)
+  /** @type {FILTransactionLoading[]} */
+  const transactionsLoading = messages.map(message => {
+    return {
+      height: message.height,
+      hash: message.cid,
+      outgoing: message.from.robust === address,
+      amount: new FilecoinNumber(message.value, 'attofil').toFil(),
+      address: message.from.robust === address
+        ? message.to.robust
+        : message.from.robust
+    }
+  })
+
+  /** @type {(FILTransaction|FILTransactionProcessing)[]} */
+  const updatedTransactions = []
+
+  for (const transactionProcessing of transactionsLoading) {
+    // Find matching transaction
+    const tx = transactions.find(tx => tx.hash === transactionProcessing.hash)
+
+    // Complete already loaded data
+    // Keep references alive by prefering `tx`
+    const transaction = tx || transactionProcessing
+
+    if (!transaction.timestamp && transaction.height) {
+      transaction.timestamp =
+        (await getTipset(transaction.height)).minTimestamp * 1000
+    }
+
+    if (
+      transaction.hash &&
+      (!transaction.status || transaction.status === 'processing') &&
+      !stateReplaysBeingFetched.has(transaction.hash)
+    ) {
+      const { hash } = transaction
+      stateReplaysBeingFetched.add(hash)
+      transaction.status = 'processing'
+      ;(async () => {
```

Comment:
Here we are starting a background task to get the tipset and the state replay, creating one task for every transaction we need to update, and making two API requests for each one.

I am concerned about the performance impact if there are more than a handful of such transactions.

The code is also mixing the implementation of the functionality we need (fetch data, transform it to the shape we need, update our internal transaction representation) with a logic to retry the requests on errors.

IMO, this design makes this code difficult to reason about and test.

For example, now that I am reading through the code more carefully again, I spotted that you are calling `updateBalance` every time you encounter a transaction that changed the state to `sent`. It's ok in the "average" case when zero or one transaction changes state, but it maybe become problematic when updating more than one transaction.

I would prefer to design the "update" functionality as follows:
- Make error statuses (e.g. we could not fetch transaction status) explicit. Store them inside the transaction instead of the current design, where the state of our code implicitly remembers them.
- Change the "update" function to be a one-pass that finishes quickly.
  1. Collect list of all transactions
  2. Find transactions we need to update
  3. Find the list of tipsets we need to fetch.
  4. Find the list of state replays we need to fetch.
  5. Fetch tipsets with controlled concurrency. 
  6. Fetch state replays with controlled concurrency. 
  7. Maybe we can do 5 & 6 serially for now (`concurrency=1`)? If the API call fails, then simply skip updating the transaction, or maybe mark the transaction status to be "could not fetch update"
  8. Update transactions using the data from 6 & 7.
  9. If at least one transaction changed the state to 'sent', then call `updateBalance` (only once).
 
- It's up to the wallet glue code to decide how often to call the "update" function - maybe more often to retry failed API calls and less often when all went well.

BTW is it safe to call `updateBalance` multiple times at the same time? I think it creates a race condition:
1. Caller A calls `updateBalance`
2. Caller B calls `updateBalance`
3. It happens that API requests finish in a different order.
4. Caller B updates the balance using the value returned by the later API call.
5. Caller A updates the balance using the value returned by the earlier API call.
6. If a transaction was committed between A & B calls, our balance shows an outdated amount.

Reply:
Updated like this <3 I also debounced `updateBalance`

> If at least one transaction changed the state to 'sent', then call updateBalance (only once).

We still need to update the balance periodically, as it can update way before a transaction gets confirmed (don't ask me how)